File: hphp/hack/src/hhbc/rx.ml
(*
 * Copyright (c) 2018, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

open Core_kernel
module Ast = Aast
module T = Aast

type conditional = bool

(* The possible Rx levels of a function or method *)
type t =
  | NonRx
  | RxLocal of conditional
  | RxShallow of conditional
  | Rx of conditional
  | Pure of conditional

let attr_is_pure ast_attr =
  snd ast_attr.Ast.ua_name = Naming_special_names.UserAttributes.uaPure

let attr_is_rx ast_attr =
  snd ast_attr.Ast.ua_name = Naming_special_names.UserAttributes.uaReactive

let attr_is_rx_shallow ast_attr =
  snd ast_attr.Ast.ua_name
  = Naming_special_names.UserAttributes.uaShallowReactive

let attr_is_rx_local ast_attr =
  snd ast_attr.Ast.ua_name = Naming_special_names.UserAttributes.uaLocalReactive

let attr_is_non_rx ast_attr =
  snd ast_attr.Ast.ua_name = Naming_special_names.UserAttributes.uaNonRx

let attr_is_rx_if_impl ast_attr =
  snd ast_attr.Ast.ua_name = Naming_special_names.UserAttributes.uaOnlyRxIfImpl

let attr_is_rx_as_args ast_attr =
  let name = snd ast_attr.Ast.ua_name in
  name = Naming_special_names.UserAttributes.uaAtMostRxAsArgs

let rx_level_from_ast ast_attrs =
  let pure = List.exists ast_attrs ~f:attr_is_pure in
  let rx = List.exists ast_attrs ~f:attr_is_rx in
  let non_rx = List.exists ast_attrs ~f:attr_is_non_rx in
  let rx_shallow = List.exists ast_attrs ~f:attr_is_rx_shallow in
  let rx_local = List.exists ast_attrs ~f:attr_is_rx_local in
  let cond =
    List.exists ast_attrs ~f:(fun a ->
        attr_is_rx_if_impl a || attr_is_rx_as_args a)
  in
  match (rx_local, rx_shallow, rx, pure) with
  | (true, false, false, false) -> Some (RxLocal cond)
  | (false, true, false, false) -> Some (RxShallow cond)
  | (false, false, true, false) -> Some (Rx cond)
  | (false, false, false, true) -> Some (Pure cond)
  | (false, false, false, false) when not cond ->
    if non_rx then
      Some NonRx
    else
      None
  | _ -> failwith "invalid combination of Rx attributes escaped the parser"

let rx_level_to_attr_string level =
  let f cond suffix =
    if cond then
      "conditional_" ^ suffix
    else
      suffix
  in
  match level with
  | NonRx -> None
  | RxLocal cond -> Some (f cond "rx_local")
  | RxShallow cond -> Some (f cond "rx_shallow")
  | Rx cond -> Some (f cond "rx")
  | Pure cond -> Some (f cond "pure")

let rx_level_from_attr_string s =
  match s with
  | "conditional_rx_local" -> Some (RxLocal true)
  | "conditional_rx_shallow" -> Some (RxShallow true)
  | "conditional_rx" -> Some (Rx true)
  | "conditional_pure" -> Some (Pure true)
  | "rx_local" -> Some (RxLocal false)
  | "rx_shallow" -> Some (RxShallow false)
  | "rx" -> Some (Rx false)
  | "pure" -> Some (Pure false)
  | _ -> None

let halves_of_is_enabled_body ast_body =
  let block = ast_body.T.fb_ast in
  match block with
  | [(_, T.If ((_, T.Id (_, const)), enabled, disabled))] ->
    if Naming_special_names.Rx.is_enabled const then
      match disabled with
      | []
      | [(_, T.Noop)] ->
        None
      | _ -> Some (enabled, disabled)
    else
      None
  | _ -> None

File: hphp/hack/src/hhbc/rx.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use naming_special_names_rust::{self as sn, user_attributes::*};
use ocamlrep_derive::{FromOcamlRep, ToOcamlRep};
use oxidized::{aast as a, ast_defs};
use std::{convert::TryFrom, str::FromStr};

pub type Conditional = bool;

/// The possible Rx levels of a function or method
#[derive(Debug, Clone, Copy, PartialEq, FromOcamlRep, ToOcamlRep)]
pub enum Level {
    NonRx,
    RxLocal,
    RxShallow,
    Rx,
    Pure,
}

#[derive(Debug, PartialEq)]
pub struct RxNone; // this is used as alternative (None was)

/// Implicitly provides via blanket impl: TryInto<&str> for Rx
impl TryFrom<Level> for &str {
    type Error = RxNone;
    fn try_from(level: Level) -> Result<Self, Self::Error> {
        use Level::*;
        match level {
            NonRx => Err(RxNone),
            RxLocal => Ok("rx_local"),
            RxShallow => Ok("rx_shallow"),
            Rx => Ok("rx"),
            Pure => Ok("pure"),
        }
    }
}

/// Implicitly provides via blanket impl: TryInto<Level> for &str
impl TryFrom<&str> for Level {
    type Error = RxNone; //&'static str;
    fn try_from(s: &str) -> Result<Self, Self::Error> {
        <Self as FromStr>::from_str(&s)
    }
}

impl FromStr for Level {
    type Err = RxNone;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        use Level::*;
        match s {
            "rx_local" => Ok(RxLocal),
            "rx_shallow" => Ok(RxShallow),
            "rx" => Ok(Rx),
            "pure" => Ok(Pure),
            _ => Err(RxNone),
        }
    }
}

impl Level {
    pub fn from_ast<Ex, Fb, En, Hi>(
        ast_attrs: impl AsRef<[a::UserAttribute<Ex, Fb, En, Hi>]>,
    ) -> Option<Self> {
        let ast_attrs = ast_attrs.as_ref();
        let attrs_contain = |name| ast_attrs.iter().any(|attr| attr.name.1 == name);
        let pure = attrs_contain(PURE);
        let rx = attrs_contain(REACTIVE);
        let non_rx = attrs_contain(NON_RX);
        let rx_shallow = attrs_contain(SHALLOW_REACTIVE);
        let rx_local = attrs_contain(LOCAL_REACTIVE);
        match (rx_local, rx_shallow, rx, pure) {
            (true, false, false, false) => Some(Self::RxLocal),
            (false, true, false, false) => Some(Self::RxShallow),
            (false, false, true, false) => Some(Self::Rx),
            (false, false, false, true) => Some(Self::Pure),
            (false, false, false, false) => {
                if non_rx {
                    Some(Self::NonRx)
                } else {
                    None
                }
            }
            _ => panic!("invalid combination of Rx attributes escaped the parser"),
        }
    }

    pub fn is_non_rx(&self) -> bool {
        if let Self::NonRx = self {
            return true;
        }
        false
    }
}

pub fn halves_of_is_enabled_body<Ex, Fb, En, Hi>(
    body: &a::FuncBody<Ex, Fb, En, Hi>,
) -> Option<(&a::Block<Ex, Fb, En, Hi>, &a::Block<Ex, Fb, En, Hi>)> {
    use a::*;
    if let [Stmt(_, Stmt_::If(if_))] = body.ast.as_slice() {
        if let (Expr(_, Expr_::Id(sid)), enabled, disabled) = &**if_ {
            let ast_defs::Id(_, name) = &**sid;
            return if name != sn::rx::IS_ENABLED {
                None
            } else {
                match disabled.as_slice() {
                    [] | [Stmt(_, Stmt_::Noop)] => None,
                    _ => Some((enabled, disabled)),
                }
            };
        }
    }
    None
}

// TODO(hrust) port remaining

#[cfg(test)]
mod tests {
    use super::*;

    use std::convert::TryInto;

    #[test]
    fn test_level_to_str_implicit_impl() {
        use Level::*;
        assert_eq!(Rx.try_into(), Ok("rx"));
        assert_eq!(RxLocal.try_into(), Ok("rx_local"));

        let s: Result<&str, RxNone> = NonRx.try_into();
        assert_eq!(s, Err(RxNone));
    }

    #[test]
    fn test_str_to_level_implicit_impl() {
        use Level::*;
        assert_eq!("rx".try_into(), Ok(Rx));
        assert_eq!("rx_shallow".try_into(), Ok(RxShallow));

        let level: Result<Level, RxNone> = "".try_into();
        assert_eq!(level, Err(RxNone));
    }
}
